modules/nf-cid/src/main/nextflow/data/cid/fs/CidPath.groovy

```diff
@@ -0,0 +1,492 @@
+    boolean startsWith(Path other) {
+        return startsWith(other.toString())
+    }
+
+    @Override
+    boolean startsWith(String other) {
+        return filePath.startsWith(other)
+    }
+
+    @Override
+    boolean endsWith(Path other) {
+        return endsWith(other.toString())
+    }
+
+    @Override
+    boolean endsWith(String other) {
+        return filePath.endsWith(other)
+    }
+
+    @Override
+    Path resolve(Path other) {
+        if( CidPath.class != other.class )
+            throw new ProviderMismatchException()
+
+        final that = (CidPath)other
+
+        if( that.fileSystem && this.fileSystem != that.fileSystem )
+            return other
+        if( that.isAbsolute() ) {
+            return that
+        } else {
+            final newPath = Path.of(filePath).resolve(that.toString())
+            return new CidPath(that.query, that.fragment, newPath.toString(), fileSystem)
+        }
+    }
+
+    @Override
+    Path resolve(String path) {
+        if( !path )
+            return this
+        final scheme = FileHelper.getUrlProtocol(path)
+        if( !scheme ) {
+            // consider the path as a cid relative path
+            return resolve(new CidPath(null,path))
+        }
+        if( scheme != SCHEME ) {
+            throw new ProviderMismatchException()
+        }
+        final that = fileSystem.provider().getPath(asUri(path))
+        return resolve(that)
+    }
+
+    @Override
+    Path relativize(Path other) {
+        if( CidPath.class != other.class ) {
+            throw new ProviderMismatchException()
+        }
+        CidPath cidOther = other as CidPath
+        if( this.isAbsolute() != cidOther.isAbsolute() )
+            throw new IllegalArgumentException("Cannot compare absolute with relative paths");
+        def path
+        if( this.isAbsolute() ) {
+            // Compare 'filePath' as absolute paths adding the root separator
+            path = Path.of(SEPARATOR + filePath).relativize(Path.of(SEPARATOR + cidOther.filePath))
+        } else {
+            // Compare 'filePath' as relative paths
+            path = Path.of(filePath).relativize(Path.of(cidOther.filePath))
+        }
+        return new CidPath(cidOther.query, cidOther.fragment, path.getNameCount()>0 ? path.toString() : SEPARATOR, null)
+    }
+
+    @Override
+    URI toUri() {
+        return asUri("${SCHEME}://${filePath}${query ? '?' + query: ''}${fragment ? '#'+ fragment : ''}")
+    }
+
+    String toUriString() {
+        return toUri().toString()
+    }
+
+    @Override
+    Path toAbsolutePath() {
+        return this
+    }
+
+    @Override
+    Path toRealPath(LinkOption... options) throws IOException {
+        return this.getTargetPath(true)
+    }
+
+    Path toTargetPath() {
+        return getTargetPath(true)
+    }
+
+    protected Path getTargetPath(boolean resultsAsPath=false){
+        return findTarget(fileSystem, filePath, resultsAsPath, parseChildrenFormFragment(fragment))
+    }
```

Comment:
@jorgee is the `resultsAsPath` parameter really needed. I've noticed that almost always invoked with `true`, except here 

https://github.com/nextflow-io/nextflow/blob/a8765e6af393811b4861ad9885e2a59d692cc8ee/modules/nf-cid/src/main/nextflow/data/cid/fs/CidFileSystemProvider.groovy#L213-L213

if needed it may be better to split in two methods with different names that clarifies the semantic

Reply:
Yes, I will change it, but not sure how to call them yet.

Reply:
The second is a the pseudo path for `#outputs` ?

Reply:
Yes, the second is for the `#outputs` or other metadata fragment that is returned as in-memory file (`CidMetadataPath`). But it also returns the target path if the key is from a `DataOutput`. I have kept `getTargetPath()` when it was `false`, and I have included `getTargetOrMetadataPath()` when it was `true`. Let me know if you find a better name.

Reply:
I think it's fine